src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1704,19 +1704,13 @@ private int tryExecute(final Runnable runnable) {
         if (TAIL_LOCK) lockTail();
         TaskNode tail = this.tail;
         TaskNode node = null;
+        int spins = 0;
         for (;;) {
             tailNext = tail.getNext();
             if (tailNext instanceof TaskNode) {
-                TaskNode tailNextTaskNode;
-                do {
-                    if (UPDATE_STATISTICS) spinMisses.increment();
-                    tailNextTaskNode = (TaskNode) tailNext;
-                    // retry
-                    tail = tailNextTaskNode;
-                    tailNext = tail.getNext();
-                } while (tailNext instanceof TaskNode);
-                // opportunistically update for the possible benefit of other threads
-                if (UPDATE_TAIL) compareAndSetTail(tail, tailNextTaskNode);
```

Comment:
I noticed this line fell off.  Do you have some definitive data that suggests it is never useful?  Is this related to the false sharing issue you observed?

Reply:
Difficult to say: I'm not aware of the original motivation of this, but looking at the logic on
```java
                if (tail.compareAndSetNext(null, node)) {
                    // try to update tail to the new node; if this CAS fails then tail already points at past the node
                    // this is because tail can only ever move forward, and the task list is always strongly connected
                    compareAndSetTail(tail, node);
                    if (TAIL_LOCK) unlockTail();
                    return EXE_OK;
                }
```
I understand that it *could* be helpful althout I see that it's enabled by walking the linked list until the end and this seems to be harmful in many test cases.
There is a specific use case which optimization was designed to help? If yes, we can run an appropriate test with/without this change to prove that's fine.

Reply:
> TBH I was hoping you found some concrete evidence. :)

Eh, is non-deterministic: the problem is that it could help only when you've already walked the list and any other thread was descheduled right before doing it, but walking the list consume CPU (on cache misses or cache coherency traffic) that could be spent by the same threads to continue their work. 
The idea is super nice from an algorithmic pov, but the risk is to always walk a "long list" before getting to help others.

Reply:
We're not walking a long list though; this method doesn't do that.  It's a single CAS operation that isn't even ever retried: if the tail is still what it was when we started walking, update it to the one node we know for sure comes after it.  If the tail was modified by someone else, give up and keep going.

Reply:
I don't think UPDATE_TAIL was enabled by default, when I tried using it I didn't see a performance improvement. I imagine it could help on machines with very few cores, but under contention when one thread submits a task, all N threads concurrently submitting work all concurrently walk the (short) list, and in the case of UPDATE_TAIL attempt a CAS that's expected to have a relatively low success rate. It's possible that we could validate the current value of `tail` equals the expected value prior to the cas, but testing this (without UPDATE_TAIL) showed a ~10% performance regression in some cases.

Reply:
> We're not walking a long list though
uh, I've now understood it, yep, you're right: so I see 3 issues with the original implementation:

- it was using a tight loop with no Thread::yield/onSpinWait 
- given that's a non-counted loop there is a chance the world will change after the `instance of` check due to the safepoint poll
- a failed compareAndSetTail could get a full memory barrier anyway (AFAIK yes)

The first problem (that was the most important one) , can be solved by doing this too:
```java
            if (tailNext instanceof TaskNode) {
                TaskNode tailNextTaskNode;
                do {
                    spins = onSpinMisses(spins);
                    tailNextTaskNode = (TaskNode) tailNext;
                    // retry
                    tail = tailNextTaskNode;
                    tailNext = tail.getNext();
                } while (tailNext instanceof TaskNode);
                // opportunistically update for the possible benefit of other threads
                if (UPDATE_TAIL) compareAndSetTail(tail, tailNextTaskNode);
            }
```
if we would like to preserve the original logic.
I suppose we would like to retest it (kindly please @carterkozak) and see if it can cause any regressions :)

Reply:
> It's possible that we could validate the current value of `tail` equals the expected value prior to the cas, but testing this (without UPDATE_TAIL) showed a ~10% performance regression in some cases.

The current code actually does this now.  I would expect that the current code is unlikely to regress significantly in performance: if the precondition is false, then the more expensive CAS and subsequent memory barrier CPU operation is not performed; if the precondition is true, then the cache line should already be hot and the CAS should be extremely unlikely to fail.

Reply:
Sorry, I mixed things up, it was QNode.compareAndSetNext that doesn't implement this check, where adding the pessimistic check reduced performance.

Reply:
let me add another thing now that I notice it :)

Reply:
I've just added the `onSpinMisses` on the queue size operations that are very likely highly contended as well with producers vs consumers.

Reply:
> Sorry, I mixed things up, it was QNode.compareAndSetNext that doesn't implement this check, where adding the pessimistic check reduced performance.

:+1: It's really about optimism.  If the CAS is sufficiently likely to succeed, the extra check is detrimental.  If the CAS is less likely to succeed (by some threshold that we don't know), then the extra check is beneficial.  I'd be interested to discover that threshold for various architectures though.

Reply:
Results using the snippet above (with UPDATE_TAIL disabled, default):
```
Benchmark                           (burstSize)  (cores)  (executorThreads)      (factory)  (spinWaitCompletion)  Mode  Cnt   Score   Error  Units
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16   EQE_NO_LOCKS                 false  avgt    4  43.368 ± 6.417  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16   EQE_NO_LOCKS                  true  avgt    4  42.429 ± 3.136  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16  EQE_SPIN_LOCK                 false  avgt    4  18.775 ± 1.027  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16  EQE_SPIN_LOCK                  true  avgt    4  25.368 ± 1.127  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200   EQE_NO_LOCKS                 false  avgt    4  15.503 ± 0.874  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200   EQE_NO_LOCKS                  true  avgt    4  13.900 ± 1.844  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200  EQE_SPIN_LOCK                 false  avgt    4  18.814 ± 0.067  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200  EQE_SPIN_LOCK                  true  avgt    4  25.156 ± 0.774  us/op
```

This should be comparable to the results from the full run yesterday (snippet copied from https://github.com/jbossas/jboss-threads/pull/72#discussion_r411039051 )
```
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16              EQE_NO_LOCKS                 false  avgt    4     19.783 ±     0.499  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16              EQE_NO_LOCKS                  true  avgt    4     17.545 ±     0.929  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16             EQE_SPIN_LOCK                 false  avgt    4     18.246 ±     0.070  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16             EQE_SPIN_LOCK                  true  avgt    4     26.530 ±     0.742  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200              EQE_NO_LOCKS                 false  avgt    4     16.698 ±     0.063  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200              EQE_NO_LOCKS                  true  avgt    4     13.976 ±     0.454  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200             EQE_SPIN_LOCK                 false  avgt    4     18.055 ±     0.127  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28            100,200             EQE_SPIN_LOCK                  true  avgt    4     25.190 ±     0.222  us/op
```

I'll give it another run with UPDATE_TAIL enabled

Reply:
@dmlloyd 
>  If the CAS is less likely to succeed
it depends by what the GC and OS decide too :O

Reply:
Results with UPDATE_TAIL enabled (limited to 16,16 since 100,200 doesn't queue tasks)

```
Benchmark                           (burstSize)  (cores)  (executorThreads)      (factory)  (spinWaitCompletion)  Mode  Cnt   Score   Error  Units
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16   EQE_NO_LOCKS                 false  avgt    4  40.560 ± 6.176  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16   EQE_NO_LOCKS                  true  avgt    4  29.825 ± 4.090  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16  EQE_SPIN_LOCK                 false  avgt    4  18.503 ± 0.506  us/op
ExecutorBenchmarks.benchmarkSubmit            1       28              16,16  EQE_SPIN_LOCK                  true  avgt    4  25.452 ± 0.994  us/op
```

I won't be able to run a full suite across core counts/batches until tonight though.

Reply:
> Results using the snippet above (with UPDATE_TAIL disabled, default):
this is this PR branch + the code on https://github.com/jbossas/jboss-threads/pull/72#discussion_r411459410? 
Or it's just https://github.com/jbossas/jboss-threads/pull/72#discussion_r411459410 on top of `master`?

Reply:
The code in https://github.com/jbossas/jboss-threads/pull/72#discussion_r411459410 applied on top of 4aa539dac449afe8436dcaf9de1c52083130b9f8 from this PR. Sorry, I should have been more specific!

Reply:
@carterkozak @dmlloyd Probably for https://github.com/jbossas/jboss-threads/pull/72/commits/c1f08fe97c75443ff04217dece0d340d65832784 we can just hint with `onSpinWait` Ideally.

Reply:
I've some numbers to share soon (statistics mostly).

Reply:
158% of the `tryExecute` we're walking a list and the average size of the list walked is of 3620 nodes: this is by running the same configuration of the test mentioned by @carterkozak 
@dmlloyd I see that the assumption that the list to be walked isn't long is not valid (I'm surprised as well).

Reply:
How are you measuring? Is it possible that measuring the number of steps causes the number of steps to be longer, by lengthening the amount of time between walking the list and updating the pointer?

Reply:
I'm using LongAdder(s) and never in the tight loop while walking the list, so it's very unlikely that it will change a lot: looking at the `perfasm` output I cannot see `LongAdder::increment/add` to cost that much under load.

Reply:
At the scales we're talking about though, it could still have an impact... maybe it would be worth it to try updating a local var instead, and update the adder on method exit?  At least, if that gives similar numbers, we could be sure that it is not likely to be having an effect.

Reply:
>  try updating a local var instead, and update the adder on method exit

I'm doing like that (with a local variable) but I'm updating the `LongAdder` at the end of the loop; according to `perfasm` the cost is negligible, but I agree that I can do it at method level too to be sure...let me try

Reply:
I was modifying the executor algorithm (still at the beginning) to use a pair of `getAndSetTail` `tail::setNextRelaxed`; but probably I will be faster to modify the LongAdder usage P

Reply:
Ah okay, I thought you were doing it every time.  But yeah any apples-to-apples comparison should help us shed light.  These things are notoriously hard to measure.

Reply:
Getting similar stats then before: probably when `this.tail` is read, (many) other producers have already appended new nodes and that loop will walk that entire list (stealing CPU) while it should have just attempt to sample the very last of it (ie this.tail with this.tail.next instanceof taskNode == false) instead of "chasing" the same trough many connected stale values (maybe keeping them alive too? I've no idea how consumers interact with the list)

Reply:
So that would seem to argue in favor of opportunistically updating the tail pointer whenever we can.

Reply:
It might be helpful to collect data describing how frequently the opportunistic update succeeds, and how often the opportunistic update fails because the new value as already been applied.

Reply:
We won't know why the update fails.  It could be for any of these reasons:

* The new value was already applied, as you say
* Another thread updated `tail` to point to a node before the one we want to point to
* Another thread updated `tail` to point to a node after the one we want to point to
* No reason (failure was spurious)

Reply:
> We won't know why the update fails...

It is the answer to my question on 
 > From my understanding if tail.compareAndSetNext(null, node) succeed, none can add new TaskNode nodes until the new tail is visible.

?

Reply:
Guys I am calling it a day, that's quite late here and I see my eyes are crossing :) will add comments tomorrow :)

Reply:
No I was answering Carter on this one: "...fails because the new value [h]as already been applied".

Reply:
So my dubt is still valid: If we save to update the tail elsewhere, there is no need to compareAndSetTail and we can use something lighter (any outOrdered is fine)

Reply:
> So my dubt is still valid: If we save to update the tail elsewhere, there is no need to compareAndSetTail and we can use something lighter (any outOrdered is fine)

Doing this would risk making `tail` visibly move backwards.  The entire algorithm would have to be re-analyzed to ensure that this is OK.

Reply:
I need to understand this I think: I don't understand in which part of the code this is supposed to happen. There are parts other then https://github.com/franz1981/jboss-threads/blob/c1f08fe97c75443ff04217dece0d340d65832784/src/main/java/org/jboss/threads/EnhancedQueueExecutor.java#L1799 that can change tail?

Reply:
I will ask on the chat or in another pr about this, sorry. I don't want to pollute the comments on this pr with OT questions :)

Reply:
I _think_ I see what's going on, but it's getting late and I may just as well have bamboozled myself.

Current state of the world using the EnhancedQueueExecutor with locks disabled, defaults otherwise:
* Thread T1 submits work, which is enqueued. it reaches 'if (tail.compareAndSetNext(null, node)) {' which succeeds and the task is added to the queue, tail still references the old value.
* Thread T2 works its way up the queue following tail task nodes. CAS to update the tail node fails because the 'tail' value in scope has advanced past eqe.tail.
* T1 tail CAS completes successfully.

At this point we're at a steady state where tail.next is a TaskNode that should be the tail. With opportunistic updates disabled, all additional nodes will be added past the tail, which doesn't catch up because each tryExecute spins through the linked list and never attempts to update tail using the correct expected value.

The tail lock protects us from this case when it's enabled.

I think similar problems can arise when threads race to update the tail for opportunistic updates because we attempt to traverse the entire list and update from original-tail to the current snapshot of the potential new tail node, allowing a wider window for failure and causing additional contention tasks submitted in parallel. The perf-asm profiler confirms that enabling update-tail doesn't substantially reduce time spent chasing the tail node.

I've opened #73 with another option to individually increment the tail one node at a time (for very short lists as we expect, this isn't much different). When this succeeds, it continues following the tail, while failures result in similar behavior to this PR. This allows opportunistic updates to be taken advantage of more quickly and avoids duplicated work walking the list. A quick sample of benchmarks showed better performance, but I'll kick off a full suite to have results ready tomorrow.

Reply:
In the scenario outlined in my last comment, I think the approach in this PR effectively spin-locks when tasks are submitted to `Executor.execute(Runnable)` until the pool has drained the queue, which makes the benchmarks look nice because everything is fast, but in practice it would produce poor behavior.

Reply:
> I think the approach in this PR effectively spin-locks when tasks are submitted to Executor.execute(Runnable) until the pool has drained the queue

I'm not sure about it; it won't await the runnable to be executed but the current producer to have moved forward the tail, that's the same behaviour in the original Michael-Scott CLQ.

Reply:
I tried applying LongAdders to the compareAndSetTail calls and found a surprisingly few successful tail updates (~2,000) and ~70,000,000 failures, so I think something isn't working the way we expect.

* With the caveat that it was late at night and I may have messed something up, I can try to recreate the scenario later in the day

Reply:
Which `compareAndSetTail `? The one to `help` a descheduled producer or the one on UPDATE_TAIL`?

Reply:
This is not surprising.  I think that when many threads are attacking the list, it is likely that they all interfere negatively with one another.  I think that it is likely that forcibly moving the list pointer one by one like you suggested is going to be the winning strategy.  Though maybe there are some other clever optimizations to be done there as well.

Reply:
Both :-)

Master currently takes tail, walks the list to the expected tail, and attempts to CAS from that value to the new expected value. If the queue is already in a state where EQE.tail doesn't match the expected value (because neither compareAndSetTail operations validate results have been updated) we end up walking the list a lot, but eventually get to the end.

Using the approach in this PR without walking the list, if we retry by re-loading `this.tail` and `this.tail->next` is a TaskNode, how do we expect `this.tail` to progress?

Reply:
> Using the approach in this PR without walking the list, if we retry by re-loading this.tail and this.tail->next is a TaskNode, how do we expect this.tail to progress?

I suggest 2 strategies:

1. don't help tail to be changed: the progress will be made when a producer end to publish the tail. As I've mentioned: it just means that after a successfull `compareAndSetNext` if the producer thread is not descheduled, the next `compareAndSetTail` will likely happen soon
2. help changing tail without walking the list (as the [CLQ](https://github.com/eagle518/jdk-source-code/blob/91b771140de051fb843af246ab826dd6ff688fe3/jdk5.0_src/j2se/src/share/classes/java/util/concurrent/ConcurrentLinkedQueue.java#L179-L196) pre-jdk 7 does), because a stale tail is already a way to walk the list

Reply:
@dmlloyd with
> I think that it is likely that forcibly moving the list pointer one by one like you suggested is going to be the winning strategy

which comment where you referring?  of who? 
Sorry I'm a bit slow to understand it: I always interpret things in multiple ways :D

Reply:
> which comment where you referring? of who?
> Sorry I'm a bit slow to understand it: I often interpret things in multiple ways :D

I was responding to Carter:

> I've opened #73 with another option to individually increment the tail one node at a time (for very short lists as we expect, this isn't much different). When this succeeds, it continues following the tail, while failures result in similar behavior to this PR. This allows opportunistic updates to be taken advantage of more quickly and avoids duplicated work walking the list. A quick sample of benchmarks showed better performance, but I'll kick off a full suite to have results ready tomorrow.

Reply:
I still don't understand why chasing `this.tail` directly won't work: will ask some help via chat on zulip (if possible) :)

Reply:
I can chat in a half hour :-)

Reply:
It would be great thanks :) I just need a case to drawn on some paper to make me clear in what this queue differ from [CLQ](https://github.com/eagle518/jdk-source-code/blob/91b771140de051fb843af246ab826dd6ff688fe3/jdk5.0_src/j2se/src/share/classes/java/util/concurrent/ConcurrentLinkedQueue.java#L179-L196)